### PR TITLE
Ruby & JS - Under the hood

### PR DESCRIPTION
**Pull request by michaeldupuisjr**
Adds post providing an overview of how Ruby code executes differently from JS.

I'm looking for feedback on technical side of this post. Specifically, I'm not sure how accurate and comprehensive the post is in covering the Virtual Machine portion of Ruby.

Thanks!

### CODE REVIEW

File: source/posts/2015-02-18-rubyists-guide-to-executing-javascript.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Again, the parser creates the Abstract Syntax Tree that the virtual machine compiles and interprets.
+
+## Tokenizer/Lexer
+If you’re wondering how Ruby knows that `print` is a separate element in the language from `'Hello, World'`, then you’re understanding the function of the Lexer or Tokenizer. The Tokenizer scans your line of Ruby code, character-by-character and determines where the "words" of the language begin and end. The Tokenizer can tell the difference between a space separating words and a space separating a method name from its arguments.
+
+And that’s the 10,000 foot lifecycle of a Ruby statement, as it goes from Tokenization to becoming machine code. If you’re looking for the microscopic explanation, I’d recommend [Ruby Under a Microscope](http://www.nostarch.com/rum).
+
+# JavaScript
+## Client-side
+Most browsers implement Just In Time (JIT) compiling. This means that the JavaScript code you write is compiled right before it gets executed by the virtual machine; though, in JavaScript, the interpreter is not referred to as a virtual machine, but as a JavaScript engine.
```
Comment (danmcclain) [inline source/posts/2015-02-18-rubyists-guide-to-executing-javascript.md]: Link to wiki on JIT?

**Comment by danmcclain** [comment]
:+1: Update for today's date

**Comment by michaeldupuisjr** [comment]
Updated.

**Comment by brzpegasus** [comment]
@michaeldupuisjr Interesting perspective starting from machine code and walking up to the more abstract language. I keep finding myself reversing the process as I'm reading through to make sure the whole picture is clearly painted. This is the order I gather from your post:

Ruby > tokenized and parsed into an AST > interpreted into C by YARV...

And then you're describing the LLVM compiling byte code to an assembly-like language. What's going on aroung the LLVM? How does C turn into byte code? And how do you go from assembly-ish to machine code? Also, might be good to specify differences between a compiler and interpreter.

File: source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+## Parser
+You can think of a node on the Abstract Syntax Tree as an atomic representation of a Ruby grammar rule. The reason that Ruby knows to print “Hello, World” when it sees `print 'Hello, World'` is because the parser knows that `print` is a method and the string `'Hello, World'` is its argument. These syntax rules are located inside of a language’s grammar rule file.
+
+Again, the parser creates the Abstract Syntax Tree that the virtual machine compiles and interprets.
+
+## Tokenizer/Lexer
+If you’re wondering how Ruby knows that `print` is a separate element in the language from `'Hello, World'`, then you’re understanding the function of the Lexer or Tokenizer. The Tokenizer scans your line of Ruby code, character-by-character and determines where the "words" of the language begin and end. The Tokenizer can tell the difference between a space separating words and a space separating a method name from its arguments.
+
+And that’s the 10,000 foot lifecycle of a Ruby statement, as it goes from Tokenization to becoming machine code. If you’re looking for the microscopic explanation, I’d recommend [Ruby Under a Microscope](http://www.nostarch.com/rum).
```
Comment (bcardarella) [inline source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md]: these parts might benefits from a diagram showing the input/output of each layer

File: source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+# JavaScript
+## Client-side
+Most browsers implement [Just-In-Time (JIT) compiling](http://en.wikipedia.org/wiki/Just-in-time_compilation). This means that the JavaScript code you write is compiled right before it gets executed by the virtual machine; though, in JavaScript, the interpreter is not referred to as a virtual machine, but as a JavaScript engine.
+
+V8 is the engine that interprets and executes JavaScript in the Chrome browser, Nitro is the engine for Safari, SpiderMonkey for Firefox, and Chakra on Internet Explorer. The efficiency with which a browser interprets JavaScript accounts for much of its performance these days, especially as JavaScript-heavy, Single Page Applications become increasingly important.
```
Comment (bcardarella) [inline source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md]: While this is mostly true the DOM is actually the biggest roadblock towards javascript performance in the browser. The DOM's javascript functions are run *synchronously* and not *asynchronously* which means that you have to wait for the previous command the finish before the next. This is why animations are so slow in browsers when compared to native

Comment (michaeldupuisjr) [inline source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md]: Ah ok. I'll add a qualifying statement around this idea.

**Comment by michaeldupuisjr** [comment]
Thanks for the feedback @brzpegasus! Exactly what I was looking for. I'm going to flesh this out a bit more.

File: source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md (hunk 2 above)
Comment (michaeldupuisjr) [inline source/posts/2015-02-23-rubyists-guide-to-executing-javascript.md]: I added a diagram just before I jump into the machine code. As @brzpegasus noted, it's a little unconventional to work backwards from the machine code, so having the diagram there should also give readers a better sense of how stuff is tying together.

**Comment by michaeldupuisjr** [comment]
Updated per Brian and Estelle's comments. @brzpegasus I hope the Virtual Machine portion of the post is clearer now. I'm still wrapping my head around the YARV and LLVM stuff myself...

**Comment by brzpegasus** [comment]
Looks good, @michaeldupuisjr!

**Comment by danmcclain** [comment]
Update the date again please

**Comment by michaeldupuisjr** [comment]
Date was updated. Should be good to go.

**Pull request merged**